File: leet_0905_sort_by_parity___narrowing.rs
impl Solution {
    pub fn sort_array_by_parity(nums: Vec<i32>) -> Vec<i32> {
        let mut nums = nums;
        let mut i = 0;
        let mut j = nums.len() - 1;
        while i < j {
            while i < j && nums[i] % 2 == 0 {
                i += 1;
            }
            while i < j && nums[j] % 2 == 1 {
                j -= 1;
            }
            nums.swap(i, j);
        }
        nums
    }
}

File: leet_0905_sort_by_parity___narrowing.go
func sortArrayByParity(nums []int) []int {
    n := len(nums)
    i := 0
    j := n - 1
    for i < j {
        for i < j && nums[i] % 2 == 0 {
            i++
        }
        for i < j && nums[j] % 2 == 1 {
            j--
        }
        nums[i], nums[j] = nums[j], nums[i]
    }
    return nums
}
